File: game.py
# Implements the game class for the 10000 dice game

from die import *
from player import *
from collections import Counter


class Game():

    GAME_LIMIT = 10000

    def __init__(self, player_list):
        self.game_won = False
        self.players = player_list
        self.current_player = 0
        self.play()
        

    def play(self):
    
        while not self.game_won:

            print("\n\nPlayer {0}'s turn".format(self.current_player + 1))
            self.p = self.players[self.current_player]
            print("============")
            self.take_turn()
            self.next_player()
            
        print("Game over! Congratulations Player {0}!".format(self.current_player + 1) )
        return
        
        
    def take_turn(self, turn_total = 0):
        # controls the rolling of dice, holding of dice, and whether or not the 
        # user would like to keep going
        turn_total = turn_total
        self.p.roll_dice()
        roll_val, scoring_dice = self.calculate_roll_value()
        print("Score for this roll: ", roll_val)
        print("Total for this turn so far: ", turn_total)
        if roll_val == 0:
            print("Bust!")
            self.show_score()
            self.p.reset_dice()
            return
            
        # made some points, what do you want to do?
        go_again = input("Roll again? (y/n): ")
        if go_again.lower() == "y":
            # IF ALL DICE ARE SCORING THEN RESET AND START AGAIN
            if scoring_dice == len(self.p.dice):
                # all dice score some point value, reset and roll all 6 again
                self.p.reset_dice()
                
            else:    
                holds = input("Do you want to hold any dice? (y/n): ")
                if holds.lower() == "y":
                    # show dice that were rolled this turn
                    for i in range(len(self.p.dice)):
                        print("D{0}".format(i + 1), end = " ")
                    print("")
                    for d in self.p.dice:
                      print(" {0}".format(d.get_value()), end = " ")
                    
                    dice_to_hold = input("\nWhich dice do you want to hold? \n" + \
                                            "Enter the number (D1 = 1): ")
                    dice_to_hold.strip()
                    holding = [int(index) - 1 for index in dice_to_hold.split(" ")]
                    
                    self.p.hold(holding)
                    turn_total += roll_val
                
                else: # didn't hold any scoring dice    
                    turn_total += 0

            self.take_turn(turn_total)

        else: # player is done, next player
            turn_total += roll_val
            if self.p.is_on_board(turn_total): 
                # only add points if player is on board, requires 1000+ points in one roll
                self.p.add_to_score(turn_total)

            self.p.reset_dice()
            if self.p.get_total_score() >= self.GAME_LIMIT:
                self.game_won = True
                print("YOU WIN!")
            self.show_score()
            return
        
    def next_player(self):
        # ensures that the game loops through player list
        if self.current_player < len(self.players) - 1:
            self.current_player += 1
        else:
            self.current_player = 0
        
       
    def calculate_roll_value(self):
        sum = 0
        scoring_dice = 0
        # see what was rolled
        rolled_dice = self.p.dice
        
        values = [d.get_value() for d in rolled_dice]
        counted_values = Counter(values)
        print("\nCOUNTED: ", counted_values)
        
        at_least_three = [d for d in counted_values if counted_values[d] > 2]
        less_than_three = [d for d in counted_values if counted_values[d] < 3]
        
        if len(less_than_three) == 6:  # rolled one of each, a straight
            sum = 1500
            scoring_dice = 6
            return sum, scoring_dice
            
        if len(less_than_three) == 3 and len(at_least_three) == 0 and len(rolled_dice) == 6:  
            # rolled all 6, only three faces showed up
            # check if three pairs
            # assume true until proved wrong
            is_three_pair = True
            for d in less_than_three:
                if counted_values[d] != 2:
                    # if one of them isn't a pair, three-pair is bust
                    is_three_pair = False
                    
            if is_three_pair: # got to here, must be true
                sum = 1500
                scoring_dice = 6
                return sum, scoring_dice
            
        # three or more of a number
        for val in at_least_three:
            if counted_values[val] > 3:
                if val == 1:  
                    #special case, 3 1's = 1000
                    sum += (1000) * (2 * (counted_values[val] - 3))
                else:  # not 1's
                    # the first 3 are still the value * 100, every addition value doubles score
                    sum += (val * 100) * (2 * (counted_values[val] - 3))
                scoring_dice += counted_values[val]
            else:
                # 1 is a special case, three 1's = 1000
                if val == 1:
                    sum += 1000
                else:
                    sum += (val * 100)
                
                scoring_dice += counted_values[val]
                    
        # one or two of a number, only 1 and 5 count
        for val in less_than_three:
            if val == 1:
                scoring_dice += counted_values[val]
                sum += 100 * counted_values[val]
            elif val == 5:
                scoring_dice += counted_values[val]
                sum += 50 * counted_values[val]
        return sum, scoring_dice
        
    def show_score(self):
        print("Player {0}'s score: ".format(self.current_player + 1), self.p.get_total_score())
        

File: java/Game.java
import java.io.*;
import java.util.*;

public class Game {
    
    boolean game_won;
    ArrayList<Player> player_list;
    int current_player;
    static int GAME_LIMIT = 10000;
    
    public Game(ArrayList<Player> p_list) {
        this.player_list = p_list;
        this.game_won = false;
        
        // an index to be used for the player_list
        this.current_player = 0;
        play();
    }
    

    public void play() {
        while (!this.game_won) {
            System.out.println("\nPlayer " + (this.current_player + 1) + "'s turn");
            Player p = this.player_list.get(this.current_player);
            System.out.println("============");
            take_turn(p);
            next_player();
        }
        System.out.println("Game over, congratulations Player " + (this.current_player + 1) + "!");
        return;
    }
    
    public void take_turn(Player p) {
        boolean is_rolling = true;
        int turn_total = 0;
        while (is_rolling) {
            p.roll_dice();
            ArrayList<Integer> roll_results = calculate_roll_value(p);
            int roll_score = roll_results.get(0);
            int scoring_dice = roll_results.get(1);

            // check for no points
            if (roll_score == 0) {
                // bust
                is_rolling = false;
                System.out.println("Bust!");
                p.get_score();
                this.show_score(p);
                p.reset_dice();
                break;
            }
            
            // player scored something, ask what to do next
            Scanner sc = new Scanner(System.in);
            System.out.print("Roll again? (y/n): ");
            String go_again = sc.next().toLowerCase();
            int y_or_n = go_again.compareTo("y");
            
            if(y_or_n == 0) { // player said yes
                if(scoring_dice == 6) {
                    // all dice score some value, reset and reroll all
                    System.out.println("resetting dice...");
                    p.reset_dice();
                }
                else {
                    // at least one didn't score, so offer to hold the ones that did
                    System.out.print("Hold any dice? (y/n): ");
                    String holding = sc.next().toLowerCase();
                    int hold_y_n = holding.compareTo("y");
                    if (hold_y_n == 0) { // yes to hold some
                        System.out.println();
                        // show dice and ask player to choose which
                        ArrayList<Die> rolled = p.get_rolled_dice();
                        for(int i = 0; i < rolled.size(); i++) {
                            System.out.print("D" + (i + 1) + " ");
                        }
                        
                        System.out.println();
                        for(Die d : rolled) {
                            System.out.print(" " + d.get_value() + " ");
                        }
                        System.out.println();
                        
                        // choose which
                        System.out.println("\nWhich dice do you want to hold?");
                        System.out.print("Enter the numbers (D1 = 1) separated only by commas: ");
                        
                        ArrayList<Integer> dice_to_hold = new ArrayList<Integer>();
                        String dice_indexes = sc.next();
                        
                        for (int i = 0; i < dice_indexes.length(); i++) {
                            String die_num = Character.toString(dice_indexes.charAt(i));
                            int is_comma = die_num.compareTo(",");
                            if (is_comma == 0) {
                                continue;
                            }
                            else {
                                // add the values to the hold list, adjusted for 0-index
                                dice_to_hold.add(Integer.parseInt(die_num) - 1);
                                p.hold(dice_to_hold);
                                turn_total += roll_score;
                            }
                        }

                        System.out.println("\n\n DICETOHOLD = " + dice_to_hold + "\n\n");
                    }
                    else {
                        // didn't hold any scoring dice, so lose their point values
                        turn_total += 0;
                    }
                }
            }
            else { // player said no to roll again
                turn_total += roll_score;
                if (p.is_on_board(turn_total)) {
                    // only add points if player is on board
                    // requires 1000+ in one roll
                    p.add_to_score(turn_total);
                }
                
                p.reset_dice();
                if (p.get_score() >= GAME_LIMIT) {
                    this.game_won = true;
                }
                this.show_score(p);
                break;
            }
        }
        return;
    }
    
    public ArrayList<Integer> calculate_roll_value(Player p) {
        ArrayList<Integer> results = new ArrayList<Integer>();
        // result will have two values, the point value and the number of scoring dice
        
        int sum = 0;
        int scoring_dice = 0;
        ArrayList<Die> rolled_dice = p.get_rolled_dice();
        
        // count occurrences
        Map<Integer, Integer> counted = new HashMap<Integer, Integer>();
        for (int i = 0; i < rolled_dice.size(); i++) {
            int die_face_val = rolled_dice.get(i).get_value();
            if(!counted.containsKey(die_face_val)) {
                // value is not yet in the map, add it
                counted.put(die_face_val, 1);
            }
            else { // value exists
                int old_count = (int) counted.get(die_face_val);
                counted.put(die_face_val, old_count + 1);
            }
        }
        
        // counted should contain something like:
        // {1=3, 4=2, 6=1}
        
        // rules vary for scoring if at least 3 of a value has been rolled
        ArrayList<Integer> at_least_three = new ArrayList<Integer>();
        ArrayList<Integer> less_than_three = new ArrayList<Integer>();
        
        for( Map.Entry<Integer, Integer> entry : counted.entrySet()) {
            int face_val = entry.getKey();
            int occurrences = entry.getValue();
            
            if (occurrences < 3) {
                less_than_three.add(face_val);
            }
            else {
                at_least_three.add(face_val);
            }
        }
        // the roll values are properly sorted between the lists
        
        // first take care of special cases
        if (less_than_three.size() == 6) {
            // rolled a straight
            sum = 1500;
            scoring_dice = 6;
            results.add(sum);
            results.add(scoring_dice);
            System.out.println("Straight");
            return results;
        }

        if (less_than_three.size() == 3 && at_least_three.size() == 0
            && rolled_dice.size() == 6) {
            // possibly rolled three pairs
            boolean is_three_pair = true;
            for(int i = 0; i < 3; i++) {
                int key = less_than_three.get(i);
                int occurrences = counted.get(key);
                if (occurrences != 2) {
                    // if one isn't a pair, three pair fails
                    is_three_pair = false;
                }
            }
            if(is_three_pair) {
                // if it got to here, then it must be three pair
                sum = 1500;
                scoring_dice = 6;
                System.out.println("Three pairs");
                results.add(sum);
                results.add(scoring_dice);
                return results;
            }
        }
        
        // special cases didn't happen, calculate point values of roll
        // three or more of a number
        for(int val : at_least_three) {
            int occurrences = counted.get(val);
            if(occurrences > 3) {
                if (val == 1) {
                    // special case, 3 1's == 1000
                    sum += 1000 * (2 * (occurrences - 3));
                }
                else { // not 1's, no special case
                    sum += (val * 100) * (2 * (occurrences - 3));   
                }
                scoring_dice += occurrences;
            }
            else { // exactly 3 of the same value rolled
                if (val == 1) {
                    sum += 1000;
                }
                else {
                    sum += (val * 100);
                }
                scoring_dice += occurrences;
            }
        }
        
        // for 1 or 2 of a number, only 1's and 5's count for points
        for (int val : less_than_three) {
            int occurrences = counted.get(val);
            if (val == 1) {
                sum += 100 * occurrences;
                scoring_dice += occurrences;
            }
            else if (val == 5) {
                sum += 50 * occurrences;
                scoring_dice += occurrences;
            }
        }
        
        System.out.println("score: " + sum);
        
        results.add(sum);
        results.add(scoring_dice);
        return results;
    }
    
    public void next_player() {
        // get next player from list, or loop back to player 1
        if(this.current_player < this.player_list.size() - 1) {
            this.current_player++;
        }
        else {
            this.current_player = 0;
        }
    }
    
    public void show_score(Player p) {
        System.out.print("Your current score is: ");
        System.out.println(p.get_score());
    }
    
    public static void main(String[] args) {
        System.out.println("Game running.");
    }
}
